### PR TITLE
CPDP-1040: enable submission of PDF with file attachments to DMS

### PR DESCRIPTION
**Pull request by mikail-khan**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala [hunk 1]
```diff
@@ -95,6 +96,17 @@ class FileUploadService(
 
   def deleteFile(envelopeId: EnvelopeId, fileId: FileId)(implicit hc: HeaderCarrier): Future[Unit] =
     fileUploadConnector.deleteFile(envelopeId, fileId)
+
+  def uploadAttachment(envelopeId: EnvelopeId, fileAttachment: FileAttachment)(
+    implicit hc: HeaderCarrier): Future[Unit] =
+    fileUploadFrontendConnector.upload(
+      envelopeId,
+      FileId("anything"),
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala]: Shouldn't `fileId` be unique?

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala [hunk 2]
```diff
@@ -58,14 +69,38 @@ class DmsSubmissionService[F[_]](
...
       _ = pdfDoc.close()
-      submission = DmsSubmissionService.createSubmission(metadata, envId, LocalDateTime.now(clock), Attachments.empty)
+      submission = DmsSubmissionService.createSubmission(metadata, envId, LocalDateTime.now(clock), 0)
       summaries = PdfAndXmlSummaries(pdfSummary)
       dmsSubmission = DmsSubmissionService.createDmsSubmission(metadata)
       _ <- fileUpload.submitEnvelope(submission, summaries, dmsSubmission)
     } yield envId
   }
 
+  override def submitPdfToDmsWithAttachments(html: String, attachments: List[FileAttachment], metaData: DmsMetadata)(
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: Implement `submitToDms` in terms of `submitPdfToDmsWithAttachments `.

File: app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala (hunk 1 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala]: Yep, I think so. It looks like the hardcoded values e.g FileId("pdf") etc are there in the scenarios where only one file needs to be sent so there is no need to generate unique values. I will just generate a random uuid string and set that as the value.

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala (hunk 2 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: I'll take a look.

Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: don't you mean the other way around? 
i.e. implement `submitPdfToDmsWithAttachments` in terms of `submitToDms`
i.e. leverage the html -> pdf functionality in `submitToDms` inside `submitPdfToDmsWithAttachments`

File: app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala (hunk 1 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/fileupload/FileUploadService.scala]: fixed

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala (hunk 2 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionAlgebra.scala]: fixed

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import play.api.mvc.{ MultipartFormData, Request }
+
+object DmsSubmissionWithAttachmentsRequestInterpreter {
+  def apply(
+    request: Request[MultipartFormData[TemporaryFile]]): Either[String, (DmsHtmlSubmission, List[FileAttachment])] = {
+
+    val maybeHtml: Option[Seq[String]] = request.body.dataParts.get("html")
+    val maybeMetadata: Option[DmsMetadata] = Json
+      .toJson(request.body.dataParts.filterKeys(_ == "html").mapValues(_.mkString("")))
+      .validate[DmsMetadata] match {
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala]: done

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      case JsSuccess(metadata: DmsMetadata, _) => Some(metadata)
+      case JsError(_)                          => None
+    }
+
+    val maybeFiles: List[FileAttachment] = request.body.files.map { file =>
+      {
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala]: done

File: app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        val filename = Paths.get(file.filename).getFileName
+        val bytes = readAllBytes(file.ref.moveTo(Paths.get(s"/tmp/dms/$filename"), replace = true))
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala]: I writing the contents to disk and then converting them to Bytes - is there a better way?

[the remark this answers is not in the document]
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/dms/DmsSubmissionWithAttachmentsRequestInterpreter.scala]: Isn't it already written on the disk? 

Isn't `readAllBytes(file.ref.path)` enough?

**Pull request merged**